Contacts only calculate fullName once and have sortable fullnames

roostr opened the pull request:
This PR enables a fix of [Signal-iOS #1487](https://github.com/WhisperSystems/Signal-iOS/issues/1487)

1. Adds caching of the calculated fullName value (which will slightly improve performance)
2. Incorporates fullNames that respect the first-name-first rules of the currently unmerged PR #22
3. Adds two new fullName properties that can be used for sorting comparators

michaelkirk commented on src/Contacts/Contact.m:
```diff
@@ -36,17 +50,54 @@ - (instancetype)initWithContactWithFirstName:(NSString *)firstName
+    return _fullName;
+}
+
+- (NSString *)comparableNameFirstLast {
```
Nice, can you add some test cases for the sorting scenarios in your comment? 

SSK tests are run in the context of ../SignalServiceKit/Example/TSKitiOSTestApp/

roostr replied:
Absolutely!

michaelkirk commented on src/Contacts/Contact.m:
```diff
@@ -3,6 +3,15 @@
 #import "TSStorageManager.h"
 
+@interface Contact ()
+{
```
How about using @synthesize instead?

roostr replied:
Yup, sure thing.

michaelkirk commented on src/Contacts/Contact.m:
```diff
@@ -19,6 +28,11 @@ - (instancetype)initWithContactWithFirstName:(NSString *)firstName
         _image                = image;
 
+        // Next three fields are generated on-demand on first access
+        _fullName                = nil;
```
not necessary to initialize ivars

roostr replied:
True. Included this to be declarative and reduce chance of ambiguity for future readers/contributors, but I'm fine with removing it in the interest of brevity.

michaelkirk reviewed (commented):
Glad to see this. Just some minor feedback

roostr commented:
All feedback has been incorporated in the new 2c1fc80 branch. Also, moved the comparator into the Contact model object for ease of testing.

michaelkirk commented on tests/Contacts/ContactSortingTest.m:
```diff
@@ -0,0 +1,119 @@
+        [contacts addObject:c];
+    }
+    
+    return [NSArray arrayWithArray:contacts]; // Return an immutable for good hygene
```
`[contacts copy]` is a more idiomatic way to return an immutable object from it's mutable variant.

roostr replied:
Sounds good, will force push a new branch with both instances changed once you complete the review.

michaelkirk reviewed (commented):
Tiny nit in the test case, but otherwise they read really clearly. Nice approach!

michaelkirk commented on src/Contacts/Contact.h:
```diff
@@ -1,5 +1,6 @@
 #import <AddressBook/AddressBook.h>
 #import <Foundation/Foundation.h>
+#import <UIKit/UIKit.h>
```
Can we get away with just forward declaring 'UIImage' instead?

roostr replied:
Yup, good idea. Will push that once you've completed the review.

roostr commented:
Pushed 157bda9, which only differs from the previous by changing the new `[NSArray arrayWithArray]` to `array copy` and forward declares `UIImage` so a `UIKit` import is not necessary.

roostr committed:
Only calculate fullName once, & sortable fullnames

1. Adds caching of the calculated fullName value
   (which will slightly improve performance)
2. Incorporates fullNames that respect the first-name-first
   rules of the currently unmerged PR #22
3. Adds two new fullName properties that can be used
   for sorting comparators
4. Move the comparator into the model object for easy testing

Includes tests to ensure that the first name first and last name
first sorts are behaving as expected.

// FREEBIE
